llpc/context/llpcCompiler.cpp

```diff
@@ -728,6 +728,221 @@ template <typename T> static FormattedBytes formatBytesLittleEndian(ArrayRef<T>
+  }
+
+  Result result = Result::Success;
+  ElfPackage candidateElf;
+
+  // Compile
+  GraphicsContext graphicsContext(m_gfxIp, pipelineInfo, &pipelineHash, &cacheHash);
+  Context *context = acquireContext();
+  context->attachPipelineContext(&graphicsContext);
+  result = buildUnlinkedShaderInternal(context, shaderInfo, stage, candidateElf, pipelineOut->stageCacheAccesses);
+  releaseContext(context);
+  if (result == Result::Success) {
+    elfBin.codeSize = candidateElf.size();
+    elfBin.pCode = candidateElf.data();
+  }
+
+  if (result == Result::Success) {
+    if (pipelineInfo->pfnOutputAlloc) {
+      void *allocBuf = pipelineInfo->pfnOutputAlloc(pipelineInfo->pInstance, pipelineInfo->pUserData, elfBin.codeSize);
+
+      uint8_t *code = static_cast<uint8_t *>(allocBuf);
+      memcpy(code, elfBin.pCode, elfBin.codeSize);
+
+      pipelineOut->pipelineBin.codeSize = elfBin.codeSize;
+      pipelineOut->pipelineBin.pCode = code;
+    } else {
+      // Allocator is not specified.
+      result = Result::ErrorInvalidPointer;
+    }
+  }
+
+  return result;
+}
+
+// Build the whole graphics pipeline. If missing elfPackage of a certain stage, we will build it firstly, and
+// link them to the full pipeline at last.
+//
+// @param [in]  pipelineInfo    Info to build this shader module
+// @param [out] pipelineOut  : Output of building this shader module
+// @param [in]  elfPackage :  Early compiled elfPackage
+// @returns : Result::Success if successful. Other return codes indicate failure.
+Result Compiler::buildGraphicsPipelineWithElf(const GraphicsPipelineBuildInfo *pipelineInfo,
+                                              GraphicsPipelineBuildOut *pipelineOut, const BinaryData *elfPackage) {
+  // clang-format off
+  SmallVector<const PipelineShaderInfo *, ShaderStageGfxCount> shaderInfo = {
+    &pipelineInfo->vs,
+    &pipelineInfo->tcs,
+    &pipelineInfo->tes,
+    &pipelineInfo->gs,
+    &pipelineInfo->fs,
+  };
+  // clang-format on
+
+  ElfPackage elf[UnlinkedStageCount] = {};
+
+  MetroHash::Hash cacheHash = {};
+  MetroHash::Hash pipelineHash = {};
+  cacheHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, true);
+  pipelineHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, true);
+
+  GraphicsContext graphicsContext(m_gfxIp, pipelineInfo, &pipelineHash, &cacheHash);
+  Context *context = acquireContext();
+  context->attachPipelineContext(&graphicsContext);
+
+  Result result = Result::Success;
+  for (unsigned idx = 0; idx < UnlinkedStageCount; idx++) {
+    if (elfPackage[idx].pCode) {
+      auto data = reinterpret_cast<const char *>(elfPackage[idx].pCode);
+      elf[idx].assign(data, data + elfPackage[idx].codeSize);
+    } else {
+      graphicsContext.setUnlinked(true);
+      result = buildUnlinkedShaderInternal(context, shaderInfo, static_cast<UnlinkedShaderStage>(idx), elf[idx],
```

Comment:
What happens when one build fails and the next one succeeds?

Reply:
Fixed.  Once it fails, we will return error code.